agent/hcp/telemetry/otel_sink.go

```diff
@@ -0,0 +1,189 @@
+package telemetry
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/agent/hcp/client"
+
+	gometrics "github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	"go.opentelemetry.io/otel/attribute"
+	otelmetric "go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/instrument"
+	otelsdk "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+type OTELSinkOpts struct {
+	Reader otelsdk.Reader
+	Logger hclog.Logger
+}
+
+type OTELSink struct {
+	spaceReplacer *strings.Replacer
+	logger        hclog.Logger
+
+	meterProvider *otelsdk.MeterProvider
+	meter         *otelmetric.Meter
+
+	gaugeStore *gaugeStore
+
+	gaugeInstruments     map[string]*instrument.Float64ObservableGauge
+	counterInstruments   map[string]*instrument.Float64Counter
+	histogramInstruments map[string]*instrument.Float64Histogram
+
+	mutex sync.Mutex
+}
+
+func NewOTELReader(client client.MetricsClient, endpoint string, exportInterval time.Duration) otelsdk.Reader {
+	exporter := NewOTELExporter(client, endpoint)
+	return otelsdk.NewPeriodicReader(exporter, otelsdk.WithInterval(exportInterval))
+}
+
+func NewOTELSink(opts *OTELSinkOpts) (*OTELSink, error) {
+	if opts.Logger == nil {
+		return nil, fmt.Errorf("failed to init OTEL sink: provide valid OTELSinkOpts Logger")
```

Comment:
Personally, I wouldn't hardcode type names in error messages:
```suggestion
		return nil, fmt.Errorf("ferror: logger cannot be nil")
```

Reply:
Yep thanks! Fixed it, will resolve but feel free to suggest changes on the messages again.